drivers/i2c/i2c_omap.c

```diff
@@ -0,0 +1,710 @@
+	/* Call a lower-level function to continue the transfer */
+	do {
+		result = i2c_omap_transfer_message_ll(dev);
+		time_left--;
+	} while (result == RETRY && time_left);
+
+	/* If no errors occurred, return success */
+	if (!result) {
+		return 0;
+	}
+
+	/* Handle timeout or specific error conditions */
+	if (result & (I2C_OMAP_STAT_ROVR | I2C_OMAP_STAT_XUDF)) {
+		i2c_omap_reset(dev);
+		i2c_omap_init_ll(dev);
+		/* Return an error code based on whether it was a timeout or buffer error */
+		return -EIO; /* Receiver overrun or transmitter underflow */
+	}
+	/* Handle arbitration loss and NACK errors */
+	if (result & (I2C_OMAP_STAT_AL | -EAGAIN)) {
+		return -EAGAIN;
+	}
+	if (result & I2C_OMAP_STAT_NACK) {
+		/* Issue a STOP condition after NACK */
+		i2c_base_addr->CON |= I2C_OMAP_CON_STP;
+		return -ENOMSG; /* Indicate a message error due to NACK */
+	}
+
+	/* Return a general I/O error if no specific error conditions matched */
+	return -EIO;
+}
+
+/**
+ * @brief Performs a common transfer operation for OMAP I2C devices.
+ *
+ * This function is responsible for transferring multiple I2C messages in a common way
+ * for OMAP I2C devices. It waits for the bus to be idle, then iterates through each
+ * message in the provided array and transfers them one by one using the i2c_omap_transfer_message()
+ * function. After all messages have been transferred, it waits for the bus to be idle again
+ * before returning.
+ *
+ * @param dev The pointer to the I2C device structure.
+ * @param msg An array of I2C messages to be transferred.
+ * @param num The number of messages in the array.
+ * @param polling Specifies whether to use polling or interrupt-based transfer.
+ * @param addr The I2C target address.
+ * @return 0 on success, or a negative error code on failure.
+ */
+static int i2c_omap_transfer_main(const struct device *dev, struct i2c_msg msg[], int num,
+				  bool polling, uint16_t addr)
+{
+	int ret;
+
+	ret = i2c_omap_wait_for_bb(dev);
+	struct i2c_omap_data *data = DEV_DATA(dev);
+
+	k_sem_take(&data->lock, K_FOREVER);
+	if (ret < 0) {
+		return ret;
+	}
+	for (int msg_idx = 0; msg_idx < num; msg_idx++) {
+		ret = i2c_omap_transfer_message(dev, &msg[msg_idx], polling, addr);
+		if (ret < 0) {
+			break;
+		}
+	}
+	k_sem_give(&data->lock);
+	i2c_omap_wait_for_bb(dev);
+	return ret;
+}
+
+/**
+ * @brief OMAP I2C transfer function using polling.
+ *
+ * This function performs the I2C transfer using the OMAP I2C controller
+ * in polling mode. It calls the common transfer function with the
+ * specified messages, number of messages, and target address.
+ *
+ * @param dev Pointer to the I2C device structure.
+ * @param msgs Array of I2C messages to be transferred.
+ * @param num_msgs Number of I2C messages in the array.
+ * @param addr Target address.
+ * @return 0 on success, negative error code on failure.
+ */
+static int __maybe_unused i2c_omap_transfer_polling(const struct device *dev, struct i2c_msg msgs[],
```

Comment:
Why `__maybe_unused`? Looks like it will always be used (at least until interrupt support is added, at that point if it is selectable then you can add this back).

Reply:
Removed it for now
Maybe I could add a property to let us choose between polling and interrupt, once I add support for it